src/metadata/metadata_passive_io.c

```diff
@@ -0,0 +1,377 @@
+	   entries_in_page */
+	end = OCF_MIN(end, ocf_metadata_collision_table_entries(cache));
+
+	for (cline = begin; cline < end; cline++) {
+		ocf_part_id_t part_id;
+
+		metadata_clear_dirty_if_invalid(cache, cline);
+		metadata_clear_valid_if_clean(cache, cline);
+
+		part_id = ocf_metadata_get_partition_id(cache, cline);
+		switch (part_id) {
+			case PARTITION_FREELIST:
+				handle_previously_invalid(cache, cline);
+				break;
+			case PARTITION_DEFAULT:
+				handle_previously_valid(cache, cline);
+				break;
+			default:
+				ocf_cache_log(cache, log_crit, "Passive update: invalid "
+						"part id for cacheline %u: %hu\n", cline, part_id);
+				ENV_BUG();
+				break;
+		}
+	}
+}
+
+static void _dec_core_stats(ocf_core_t core)
+{
+	ocf_part_id_t part = PARTITION_DEFAULT;
+
+	ENV_BUG_ON(env_atomic_dec_return(&core->runtime_meta->cached_clines) < 0);
+	ENV_BUG_ON(env_atomic_dec_return(
+				&core->runtime_meta->part_counters[part].cached_clines) < 0);
+
+	ENV_BUG_ON(env_atomic_dec_return(&core->runtime_meta->dirty_clines) < 0);
+	ENV_BUG_ON(env_atomic_dec_return(
+				&core->runtime_meta->part_counters[part].dirty_clines) < 0);
+}
+
+static void cleanup_old_mapping(ocf_cache_t cache, ocf_cache_line_t begin,
+		ocf_cache_line_t end)
+{
+	ocf_cache_line_t cline;
+	uint32_t lock_idx = ocf_metadata_concurrency_next_idx(cache->mngt_queue);
+
+	/* The last page of collision section may contain fewer entries than
+	   entries_in_page */
+	end = OCF_MIN(end, ocf_metadata_collision_table_entries(cache));
+
+	for (cline = begin; cline < end; cline++) {
+		ocf_core_id_t core_id;
+		uint64_t core_line;
+		ocf_core_t core;
+
+		ocf_metadata_get_core_info(cache, cline, &core_id, &core_line);
+
+		ENV_BUG_ON(core_id > OCF_CORE_ID_INVALID);
+
+		core = ocf_cache_get_core(cache, core_id);
+		if (!core)
+			continue;
+
+		_dec_core_stats(core);
+
+		ocf_hb_cline_prot_lock_wr(&cache->metadata.lock, lock_idx, core_id,
+				core_line);
+		ocf_metadata_remove_from_collision(cache, cline, PARTITION_DEFAULT);
+		ocf_hb_cline_prot_unlock_wr(&cache->metadata.lock, lock_idx, core_id,
+				core_line);
+	}
+}
+
+static int passive_io_resume(struct ocf_request *req)
+{
+	ocf_cache_t cache = req->cache;
+	struct ocf_metadata_ctrl *ctrl = cache->metadata.priv;
+	struct ocf_io *io = (struct ocf_io*) req->data;
+	ctx_data_t *data = ocf_io_get_data(io);
+	uint64_t io_start_page = BYTES_TO_PAGES(io->addr);
+	uint64_t io_end_page = io_start_page + BYTES_TO_PAGES(io->bytes);
+	ocf_end_io_t io_cmpl = req->master_io_req;
+	enum ocf_metadata_segment_id update_segments[] = {
+		metadata_segment_sb_config,
+		metadata_segment_part_config,
+		metadata_segment_core_config,
+		metadata_segment_core_uuid,
+		metadata_segment_collision,
+	};
+	int i;
+
+	for (i = 0; i < ARRAY_SIZE(update_segments); i++) {
+		ocf_cache_line_t cache_line_range_end, cache_line_range_start;
+		enum ocf_metadata_segment_id seg = update_segments[i];
```

Comment:
The reason why cache lines range is calculated here, not in `cleanup_old_mapping()` and `update_list_segment()` is that I wanted to avoid passing `struct ocf_metadata_raw` to them. With the current implementation both functions operate only in a domain of cache lines and I prefer not to change it. I will move calculating the proper size of the range to the main function and pass it instead of end of range